grub-core/loader/multiboot_mbi2.c

```diff
@@ -390,6 +500,46 @@ grub_multiboot2_load (grub_file_t file, const char *filename)
     err = grub_multiboot2_set_console (GRUB_MULTIBOOT2_CONSOLE_EGA_TEXT,
 				       accepted_consoles,
 				       0, 0, 0, console_required);
+
+  if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+    {
+      slparams->mle_start = load_base_addr;
+      slparams->mle_size = load_size;
+
+      /* Update the MLE header. */
+      mle_header = (struct grub_txt_mle_header *)(grub_addr_t) (slparams->mle_start + slparams->mle_header_offset);
+      mle_header->cmdline_start = cmdline;
+      mle_header->cmdline_end = cmdline + cmdline_size;
+
+      if (grub_relocator_alloc_chunk_align (relocator, &ch, 0x1000000,
+					    0xffffffff - GRUB_SLAUNCH_TPM_EVT_LOG_SIZE,
+					    GRUB_SLAUNCH_TPM_EVT_LOG_SIZE, GRUB_PAGE_SIZE,
+					    GRUB_RELOCATOR_PREFERENCE_NONE, 1))
+	{
+	  grub_free (mld.buffer);
+	  return grub_error (GRUB_ERR_OUT_OF_MEMORY, "Could not allocate TPM event log area");
+	}
+
+	slparams->tpm_evt_log_base = get_physical_target_address (ch);
+	slparams->tpm_evt_log_size = GRUB_SLAUNCH_TPM_EVT_LOG_SIZE;
+
+	grub_memset (get_virtual_current_address (ch), 0, slparams->tpm_evt_log_size);
+
+	grub_dprintf ("multiboot_loader", "tpm_evt_log_base = %lx, tpm_evt_log_size = %x\n",
+		      (unsigned long) slparams->tpm_evt_log_base,
+		      (unsigned) slparams->tpm_evt_log_size);
+
+	if (grub_relocator_alloc_chunk_align (relocator, &ch, 0x1000000,
+						    0xffffffff & ~slparams->mle_ptab_size,
```

Comment:
Is this a fancy optimization or protection against underflow?

Reply:
Idea copied from Linux implementation

Reply:
Outdated